Branch-SDK/src/io/branch/referral/builders/BranchShortUrlBuilder.java

```diff
@@ -0,0 +1,212 @@
+            tags_ = new ArrayList<String>();
+        }
+        this.tags_.add(tag);
+        return this;
+    }
+
+    /**
+     * <p>Set the number of times the link should perform deep link.</p>
+     *
+     * @param type {@link int} that can be used for scenarios where you want the link to
+     *             only deep link the first time.
+     * @returnA {@link BranchShortUrlBuilder} instance.
+     */
+    public BranchShortUrlBuilder setType(int type) {
+        this.type_ = type;
+        return this;
+    }
+
+    /**
+     * <p>Adds the the given key value pair to the parameters associated with this link.</p>
+     *
+     * @param key   A {@link String} with value of key for the parameter
+     * @param value A {@link String} with value of value for the parameter
+     * @returnA {@link BranchShortUrlBuilder} instance.
+     */
+    public BranchShortUrlBuilder addParameters(String key, String value) {
+        try {
+            if (this.params_ == null) {
+                this.params_ = new JSONObject();
+            }
+            this.params_.put(key, value);
+        } catch (JSONException ignore) {
+
+        }
+        return this;
+    }
+
+    ///------------------------- link Build methods---------------------------///
+
+    /**
+     * <p>Configures and requests a short URL to be generated by the Branch servers, via a synchronous
+     * call; with a duration specified within which an app session should be matched to the link.</p>
+     *
+     * @return A {@link String} containing the resulting short URL. Null is returned in case of an error or if Branch is not initialised.
+     */
+    public String getShortUrl() {
+        String shortUrl = null;
+        if (branchReferral_ != null) {
+            shortUrl = branchReferral_.generateShortLink(alias_, type_, duration_, tags_, channel_, feature_, stage_, stringifyParams(params_), null, false);
+        }
+        return shortUrl;
+    }
+
+    /**
+     * <p>Configures and requests a short URL to be generated by the Branch servers, via a asynchronous
+     * call; The {@link Branch.BranchLinkCreateListener} is called back with the url when the url is generated.</p>
+     *
+     * @param callback A {@link Branch.BranchLinkCreateListener} callback instance that will trigger
+     */
+
+    public void generateShortUrl(Branch.BranchLinkCreateListener callback) {
+        String shortUrl = null;
+        if (branchReferral_ != null) {
+            branchReferral_.generateShortLink(alias_, type_, duration_, tags_, channel_, feature_, stage_, stringifyParams(params_), callback, true);
+        }
+    }
+
+    //------------------------- Utility methods -----------------------//
+
+    protected static String stringifyParams(JSONObject params) {
+        if (params == null) {
+            params = new JSONObject();
+        }
+        try {
+            params.put("source", "android");
+        } catch (JSONException e) {
+            e.printStackTrace();
```

Comment:
We should probably do something nicer than print stack trace (or log a message with it, as well as the stack trace, via `Log.e`).